File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # have a current max
        # have current max path
        # take the max of the left path plus the max of the right path
        # set it as current max if greater
        # when returning up to parent make sure to only return the path
        res = 0

        def dfs(node):
            nonlocal res

            if not node:
                return 0
            
            right = dfs(node.right)
            left = dfs(node.left)
            # this checks if going down right and left subtrees 
            # have a greater combined path then current max
            res = max(res, left + right)

            # return only the longest of two paths
            # can't count a node twice
            return 1 + max(left, right)            
            
        dfs(root)
        return res

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {  
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if((root == null && subRoot != null) || subRoot == null) return false;
        if(equalsTree(root,subRoot) || isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot)) return true;
         return false;

    }

    public boolean equalsTree(TreeNode root1, TreeNode root2) {
        if(root1 == null && root2 == null) return true;
        if(root1 == null && root2 != null || root1 != null && root2 == null) return false;
        if(root1.val == root2.val && equalsTree(root1.left, root2.left) 
        && equalsTree(root1.right, root2.right)) return true;

        return false;
    }
}
